doc/local_context_guidelines.md

```diff
@@ -0,0 +1,139 @@
+
+- **scopes**
+
+`scopes` is an array of string. It defines where can this local context value been referenced. The value could be a command group or a command, for example `['vm', 'network']`. If one argument can be referenced in all the commands, you can define it as `['all']`. `scopes` is meaningful only when `SET` is in actions.
+    
+- **actions**
+
+`actions` is an array of [`LocalContextAction`](#Local context action). It defines whether set the command parameter value to local context or get value from local context. We can define both `GET` and `SET` action for one argument.
+
+
+## Local context action
+
+Local context action is defined as an enum. The available values are `GET` and `SET`.
+
+- **GET**
+
+Used for retrieving value local context.
+
+- **SET**
+
+Used for saving value to local context.
+
+
+## Parameters which support `all` scope by default
+
+We define local context attribute for some command parameters by default. For commands which use them as argument name in the function signature will support local context by default.
+
+All these parameters are list here:
+
+- *resource_group_name*
+```python
+    local_context_attribute=LocalContextAttribute(
+        name='resource_group_name',
+        actions=[LocalContextAction.SET, LocalContextAction.GET],
+        scopes=[ALL]
+    ))
+```
+
+- *vnet_name*
+```python
+    local_context_attribute=LocalContextAttribute(name='vnet_name', actions=[LocalContextAction.GET])
+```
+
+- *subnet*
+```python
+    local_context_attribute=LocalContextAttribute(name='subnet_name', actions=[LocalContextAction.GET])
+```
+
+For example, below is the definition of custom function for `az network vnet create`:
+```python
+def create_vnet(cmd, resource_group_name, vnet_name, vnet_prefixes='10.0.0.0/16',
+                subnet_name=None, subnet_prefix=None, dns_servers=None,
+                location=None, tags=None, vm_protection=None, ddos_protection=None,
+                ddos_protection_plan=None):
+```
+As `resource_group_name` is defined in the signature, it will automatically support local context.
+
+
+## Suggestions
+
+- Only enable local context feature for resource dependency parameters.
+- Do not define `GET` action for `name` parameter in `create` and `delete` command.
+- Define `SET` action only in `create` command.
+
+## Examples
+In order to create a webapp, user need to prepare an appservice plan first. Previously the user need to run below commands to complete this scenario:
+```bash
+az group create --name myResourceGroup --location westus
+az appservice plan create -g myResourceGroup --name myPlan
+az webapp create -g myResourceGroup --plan myPlan --name myWebapp
+```
+The user has to input resource group name 3 times and plan name 2 times for this scenario. To enable local context feature, we need to define local context attribute as below:
```

Comment:
How about now? How many times are needed when local context is enabled. Try to make a comparison.